File: backend/routes/download.js
import express from 'express';
import fs from 'fs/promises';
import path from 'path';
import { v4 as uuidv4 } from 'uuid';
import { runQuery, getQuery, allQuery } from '../database.js';
import { YouTubeService, YouTubeDownloadError } from '../services/youtube.js';
import { AudioConverter, ConversionError } from '../services/converter.js';
import { rateLimit } from '../middleware/rateLimit.js';
import config from '../config.js';

const router = express.Router();
const youtubeService = new YouTubeService();

// Helper function to validate YouTube URL
function isValidYouTubeUrl(url) {
  return youtubeService.validateUrl(url);
}

// Process download in background
async function processDownload(downloadId, url, quality) {
  try {
    // Update status to processing
    await runQuery(
      `UPDATE downloads SET status = 'processing' WHERE id = ?`,
      [downloadId]
    );

    // Download audio
    const result = await youtubeService.downloadAudio(url, quality);

    // Update download record with success
    const now = new Date().toISOString();
    await runQuery(
      `UPDATE downloads SET
        file_path = ?,
        file_size = ?,
        status = 'completed',
        completed_at = ?
       WHERE id = ?`,
      [result.filePath, result.fileSize, now, downloadId]
    );
  } catch (error) {
    // Update download record with error
    const errorMessage = error.message || 'Unknown error occurred';
    await runQuery(
      `UPDATE downloads SET status = 'failed', error_message = ? WHERE id = ?`,
      [errorMessage, downloadId]
    );

    console.error(`Error processing download ${downloadId}:`, error);
  }
}

// POST /api/download - Create a new download
router.post('/download', rateLimit, async (req, res) => {
  try {
    const { url, quality = config.DEFAULT_QUALITY } = req.body;

    // Validation
    if (!url || !url.trim()) {
      return res.status(400).json({
        success: false,
        message: 'YouTube URL is required',
      });
    }

    if (url.length > 500) {
      return res.status(400).json({
        success: false,
        message: 'URL is too long',
      });
    }

    // Validate quality
    if (!config.ALLOWED_QUALITIES.includes(quality)) {
      return res.status(400).json({
        success: false,
        message: `Invalid quality. Allowed: ${config.ALLOWED_QUALITIES.join(', ')}`,
      });
    }

    // Validate YouTube URL
    if (!isValidYouTubeUrl(url)) {
      return res.status(400).json({
        success: false,
        message: 'Invalid YouTube URL',
      });
    }

    // Get video info
    let videoInfo;
    try {
      videoInfo = await youtubeService.getVideoInfo(url);
    } catch (error) {
      return res.status(400).json({
        success: false,
        message: error.message,
      });
    }

    // Create download record
    const result = await runQuery(
      `INSERT INTO downloads (youtube_url, title, quality, status)
       VALUES (?, ?, ?, 'pending')`,
      [url, videoInfo.title, quality]
    );

    const downloadId = result.id;

    // Start background processing
    processDownload(downloadId, url, quality).catch((error) => {
      console.error('Error in background processing:', error);
    });

    return res.status(202).json({
      success: true,
      download_id: downloadId,
      status: 'pending',
      message: 'Download queued successfully',
    });
  } catch (error) {
    console.error('Error creating download:', error);
    return res.status(500).json({
      success: false,
      message: `Error creating download: ${error.message}`,
    });
  }
});

// GET /api/download/:id - Get download status
router.get('/download/:id', async (req, res) => {
  try {
    const { id } = req.params;

    const download = await getQuery(
      `SELECT * FROM downloads WHERE id = ?`,
      [id]
    );

    if (!download) {
      return res.status(404).json({
        success: false,
        message: 'Download not found',
      });
    }

    // Calculate progress percentage
    let progress = 0;
    if (download.status === 'pending') {
      progress = 0;
    } else if (download.status === 'processing') {
      progress = 50;
    } else if (download.status === 'completed') {
      progress = 100;
    } else if (download.status === 'failed') {
      progress = 0;
    }

    return res.json({
      success: true,
      download_id: download.id,
      title: download.title,
      quality: download.quality,
      status: download.status,
      progress_percentage: progress,
      error_message: download.error_message,
      file_size: download.file_size,
      created_at: download.created_at,
      completed_at: download.completed_at,
    });
  } catch (error) {
    console.error('Error fetching download status:', error);
    return res.status(500).json({
      success: false,
      message: `Error fetching download status: ${error.message}`,
    });
  }
});

// GET /api/download/:id/file - Download the MP3 file
router.get('/download/:id/file', async (req, res) => {
  try {
    const { id } = req.params;

    const download = await getQuery(
      `SELECT * FROM downloads WHERE id = ?`,
      [id]
    );

    if (!download) {
      return res.status(404).json({
        success: false,
        message: 'Download not found',
      });
    }

    if (download.status !== 'completed') {
      return res.status(400).json({
        success: false,
        message: `Cannot download. Status: ${download.status}`,
      });
    }

    const filePath = download.file_path;

    if (!filePath) {
      return res.status(404).json({
        success: false,
        message: 'File path not found',
      });
    }

    // Check if file exists
    const exists = await fs.stat(filePath).then(() => true).catch(() => false);

    if (!exists) {
      return res.status(404).json({
        success: false,
        message: 'File not found',
      });
    }

    // Send file
    const filename = `${download.title}.mp3`;
    res.download(filePath, filename, (err) => {
      if (err) {
        console.error('Error sending file:', err);
      }
    });
  } catch (error) {
    console.error('Error downloading file:', error);
    return res.status(500).json({
      success: false,
      message: `Error downloading file: ${error.message}`,
    });
  }
});

// DELETE /api/download/:id - Delete a download
router.delete('/download/:id', async (req, res) => {
  try {
    const { id } = req.params;

    const download = await getQuery(
      `SELECT * FROM downloads WHERE id = ?`,
      [id]
    );

    if (!download) {
      return res.status(404).json({
        success: false,
        message: 'Download not found',
      });
    }

    // Delete file if exists
    if (download.file_path) {
      try {
        const exists = await fs.stat(download.file_path).then(() => true).catch(() => false);
        if (exists) {
          await fs.unlink(download.file_path);
        }
      } catch (error) {
        console.error('Error deleting file:', error);
      }
    }

    // Delete database record
    await runQuery(
      `DELETE FROM downloads WHERE id = ?`,
      [id]
    );

    return res.json({
      success: true,
      message: 'Download deleted successfully',
    });
  } catch (error) {
    console.error('Error deleting download:', error);
    return res.status(500).json({
      success: false,
      message: `Error deleting download: ${error.message}`,
    });
  }
});

export default router;

File: backend/routes/download.py
from flask import Blueprint, request, jsonify, send_file
import os
import threading
import uuid
from datetime import datetime
import database
from services.youtube import YouTubeService, YouTubeDownloadError
from middleware.rate_limit import rate_limit
import config

download_bp = Blueprint('download', __name__)
youtube_service = YouTubeService()


def process_download(download_id, url, quality):
    """Process download in background."""
    try:
        # Update status to processing
        database.run_query(
            'UPDATE downloads SET status = "processing" WHERE id = ?',
            [download_id]
        )

        # Download audio
        result = youtube_service.download_audio(url, quality)

        # Update download record with success
        now = datetime.now().isoformat()
        database.run_query(
            '''UPDATE downloads SET
                file_path = ?,
                file_size = ?,
                status = "completed",
                completed_at = ?
            WHERE id = ?''',
            [result['filePath'], result['fileSize'], now, download_id]
        )
    except Exception as error:
        # Update download record with error
        error_message = str(error) if str(error) else 'Unknown error occurred'
        database.run_query(
            'UPDATE downloads SET status = "failed", error_message = ? WHERE id = ?',
            [error_message, download_id]
        )

        print(f'Error processing download {download_id}: {error}')


@download_bp.route('/download', methods=['POST'])
@rate_limit
def create_download():
    """POST /api/download - Create a new download."""
    try:
        print('[DOWNLOAD] Request received')
        data = request.get_json()
        print(f'[DOWNLOAD] Request data: {data}')
        
        url = data.get('url', '').strip() if data else ''
        quality = data.get('quality', config.DEFAULT_QUALITY) if data else config.DEFAULT_QUALITY
        
        print(f'[DOWNLOAD] URL: {url}, Quality: {quality}')

        # Validation
        if not url:
            print('[DOWNLOAD] FAIL: URL is required')
            return jsonify({
                'success': False,
                'message': 'YouTube URL is required',
            }), 400

        if len(url) > 500:
            print('[DOWNLOAD] FAIL: URL too long')
            return jsonify({
                'success': False,
                'message': 'URL is too long',
            }), 400

        # Validate quality
        print(f'[DOWNLOAD] Allowed qualities: {config.ALLOWED_QUALITIES}')
        if quality not in config.ALLOWED_QUALITIES:
            print(f'[DOWNLOAD] FAIL: Invalid quality {quality}')
            return jsonify({
                'success': False,
                'message': f'Invalid quality. Allowed: {", ".join(config.ALLOWED_QUALITIES)}',
            }), 400

        # Validate YouTube URL
        url_valid = youtube_service.validate_url(url)
        print(f'[DOWNLOAD] URL valid: {url_valid}')
        if not url_valid:
            print('[DOWNLOAD] FAIL: Invalid YouTube URL')
            return jsonify({
                'success': False,
                'message': 'Invalid YouTube URL',
            }), 400

        # Get video info
        try:
            print('[DOWNLOAD] Fetching video info...')
            video_info = youtube_service.get_video_info(url)
            print(f'[DOWNLOAD] Video info: {video_info}')
        except YouTubeDownloadError as error:
            print(f'[DOWNLOAD] FAIL: Video info error: {error}')
            return jsonify({
                'success': False,
                'message': str(error),
            }), 400

        # Create download record
        result = database.run_query(
            'INSERT INTO downloads (youtube_url, title, quality, status) VALUES (?, ?, ?, "pending")',
            [url, video_info['title'], quality]
        )

        download_id = result['id']

        # Start background processing
        thread = threading.Thread(
            target=process_download,
            args=(download_id, url, quality),
            daemon=True
        )
        thread.start()

        return jsonify({
            'success': True,
            'download_id': download_id,
            'status': 'pending',
            'message': 'Download queued successfully',
        }), 202

    except Exception as error:
        print(f'Error creating download: {error}')
        return jsonify({
            'success': False,
            'message': f'Error creating download: {str(error)}',
        }), 500


@download_bp.route('/download/<int:download_id>', methods=['GET'])
def get_download_status(download_id):
    """GET /api/download/<id> - Get download status."""
    try:
        download = database.get_query(
            'SELECT * FROM downloads WHERE id = ?',
            [download_id]
        )

        if not download:
            return jsonify({
                'success': False,
                'message': 'Download not found',
            }), 404

        # Calculate progress percentage
        progress = 0
        if download.get('status') == 'pending':
            progress = 0
        elif download.get('status') == 'processing':
            progress = 50
        elif download.get('status') == 'completed':
            progress = 100
        elif download.get('status') == 'failed':
            progress = 0

        return jsonify({
            'success': True,
            'download_id': download.get('id'),
            'title': download.get('title'),
            'quality': download.get('quality'),
            'status': download.get('status'),
            'progress_percentage': progress,
            'error_message': download.get('error_message'),
            'file_size': download.get('file_size'),
            'created_at': download.get('created_at'),
            'completed_at': download.get('completed_at'),
        }), 200

    except Exception as error:
        print(f'Error fetching download status: {error}')
        return jsonify({
            'success': False,
            'message': f'Error fetching download status: {str(error)}',
        }), 500


@download_bp.route('/download/<int:download_id>/file', methods=['GET'])
def download_file(download_id):
    """GET /api/download/<id>/file - Download the MP3 file."""
    try:
        download = database.get_query(
            'SELECT * FROM downloads WHERE id = ?',
            [download_id]
        )

        if not download:
            return jsonify({
                'success': False,
                'message': 'Download not found',
            }), 404

        if download.get('status') != 'completed':
            return jsonify({
                'success': False,
                'message': f'Cannot download. Status: {download.get("status")}',
            }), 400

        file_path = download.get('file_path')

        if not file_path:
            return jsonify({
                'success': False,
                'message': 'File path not found',
            }), 404

        # Check if file exists
        if not os.path.isfile(file_path):
            return jsonify({
                'success': False,
                'message': 'File not found',
            }), 404

        # Send file
        return send_file(
            file_path,
            as_attachment=True,
            download_name=f'{download.get("title")}.mp3'
        )

    except Exception as error:
        print(f'Error downloading file: {error}')
        return jsonify({
            'success': False,
            'message': f'Error downloading file: {str(error)}',
        }), 500


@download_bp.route('/download/<int:download_id>', methods=['DELETE'])
def delete_download(download_id):
    """DELETE /api/download/<id> - Delete a download."""
    try:
        download = database.get_query(
            'SELECT * FROM downloads WHERE id = ?',
            [download_id]
        )

        if not download:
            return jsonify({
                'success': False,
                'message': 'Download not found',
            }), 404

        # Delete file if exists
        if download.get('file_path'):
            try:
                if os.path.isfile(download['file_path']):
                    os.unlink(download['file_path'])
            except Exception as error:
                print(f'Error deleting file: {error}')

        # Delete database record
        database.run_query(
            'DELETE FROM downloads WHERE id = ?',
            [download_id]
        )

        return jsonify({
            'success': True,
            'message': 'Download deleted successfully',
        }), 200

    except Exception as error:
        print(f'Error deleting download: {error}')
        return jsonify({
            'success': False,
            'message': f'Error deleting download: {str(error)}',
        }), 500
